Sources/Commands/Workspace.swift

```diff
@@ -89,20 +92,44 @@ public class Workspace {
         /// The current version of the dependency, if known.
         public let currentVersion: Version?
 
+        /// The dependency is in editable state i.e. user is expected to modify the sources of the dependency.
+        /// The version of the dependency will not be considered during dependency resolution.
+        var isInEditableState: Bool {
+            return basedOn != nil
+        }
+
+        /// A dependency which in editable state is based on a dependency from which it edited from.
+        /// This information is useful so it can be restored when users unedit a package.
+        let basedOn: ManagedDependency?
+
         /// The current revision of the dependency.
         ///
         /// This should always be a revision corresponding to the version in the
         /// repository, but in certain circumstances it may not be the *current*
         /// one (e.g., if this data is accessed with a different version of the
         /// package manager, which would cause an alternate version to be
         /// resolved).
+        /// For a dependency in edit mode, this will be the version at which the
+        /// the dependency was made editable.
+        // FIXME: Maybe rename to lastKnownRevision?
         public let currentRevision: Revision
 
         fileprivate init(repository: RepositorySpecifier, subpath: RelativePath, currentVersion: Version?, currentRevision: Revision) {
             self.repository = repository
             self.subpath = subpath
             self.currentVersion = currentVersion
             self.currentRevision = currentRevision
+            self.basedOn = nil
+        }
+
+        /// Create an editable managed dependency based on a dependency which was *not* in edit state.
+        fileprivate init(basedOn dependency: ManagedDependency, subpath: RelativePath, currentRevision: Revision) {
```

Comment:
I don't understand, I'm just saying instead of having the client create a ManagedDependency, have a method which returns a new ManagedDependency which is based on an existing one.